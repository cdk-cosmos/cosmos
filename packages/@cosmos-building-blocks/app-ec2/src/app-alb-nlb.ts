/* eslint-disable @typescript-eslint/camelcase */
import { CfnOutput, Construct, Fn } from '@aws-cdk/core';
import {
  CfnVPCEndpoint,
  CfnVPCEndpointService,
  IInterfaceVpcEndpoint,
  InterfaceVpcEndpoint,
  IVpc,
  IVpcEndpointServiceLoadBalancer,
  SecurityGroup,
  SubnetType,
  VpcEndpointService,
} from '@aws-cdk/aws-ec2';
import {
  NetworkLoadBalancer,
  NetworkListener,
  NetworkTargetGroup,
  ApplicationLoadBalancer,
  CfnListener,
  TargetType,
} from '@aws-cdk/aws-elasticloadbalancingv2';
import { Rule, Schedule, IRuleTarget, RuleTargetInput, IRule } from '@aws-cdk/aws-events';
import { ArnPrincipal } from '@aws-cdk/aws-iam';
import { IFunction } from '@aws-cdk/aws-lambda';
import { CfnRecordSet } from '@aws-cdk/aws-route53';
import { Bucket } from '@aws-cdk/aws-s3';
import { StringParameter } from '@aws-cdk/aws-ssm';

export interface AppProps {
  appAlbNlbVpc: IVpc;
  envName: string;
}

export interface AppAlbProps extends AppProps {
  appAlbNlbVpc: IVpc;
  albSecGroupId: string;
  bucketPrefix: string;
  certificateArn: string;
  envName: string;
  logBucket: string;
}

export interface AppNlbComboCloudwatchLambdaRuleProps {
  lambdaForUpdatingTargets: IFunction;
  scheduleRate?: string;
  // targetInputText: string;
  // eslint unhappy about any but that is the type the cdk expects
  // eslint-disable-next-line
  targetInputObject: any;
}

export interface AppNlbComboInterfaceVpcEndpointProps {
  envName: string;
  exportName?: string;
  port?: number;
  subnetType?: SubnetType;
  vpc: IVpc;
  vpcEndpointServiceLoadBalancers: IVpcEndpointServiceLoadBalancer[];
  whitelistedPrincipals?: ArnPrincipal[];
}

export interface NlbEndpointComboProps extends AppProps {
  albDnsAddress: string;
  lambdaForUpdatingTargets: IFunction;
  nlbEndpointVpc?: IVpc;
  parameterPath?: string;
  publicHostedZoneId?: string;
  recordName?: string;
  recordSetPrefix?: string;
  zoneName?: string;
}

export class AppNlbComboCloudwatchLambdaRule extends Construct {
  public readonly rule: IRule;

  constructor(scope: Construct, id: string, props: AppNlbComboCloudwatchLambdaRuleProps) {
    super(scope, id);

    const t1: IRuleTarget = {
      bind: () => ({
        id: '',
        arn: props.lambdaForUpdatingTargets.functionArn,
        input: RuleTargetInput.fromObject(props.targetInputObject),
      }),
    };

    const scheduleRate = props.scheduleRate || 'rate(180 minutes)';

    this.rule = new Rule(this, id, {
      // Name
      // Description
      schedule: Schedule.expression(scheduleRate),
      targets: [t1],
    });
  }
}

export class AppAlb extends Construct {
  public readonly albDnsAddress: string;
  public readonly listener: CfnListener;
  public readonly internalListener: CfnListener;

  constructor(scope: Construct, id: string, props: AppAlbProps) {
    super(scope, id);

    const vpc = props.appAlbNlbVpc;

    const albSecurityGroup = SecurityGroup.fromSecurityGroupId(this, 'sg', props.albSecGroupId);

    // Application Load Balancers
    const uiAlb = new ApplicationLoadBalancer(this, `${props.envName}-alb`, {
      vpc,
      internetFacing: false,
      loadBalancerName: `${props.envName}-alb`,
      deletionProtection: false,
      // vpcSubnets: subnets,
      securityGroup: albSecurityGroup,
    });

    const logBucket = Bucket.fromBucketName(this, 'logBucket', props.logBucket);
    uiAlb.logAccessLogs(logBucket, props.bucketPrefix);

    // const action = new CfnListener.ActionProperty();
    const listener = new CfnListener(this, 'listener', {
      defaultActions: [
        {
          fixedResponseConfig: {
            contentType: 'text/plain',
            messageBody: 'hi from the default last rule',
            statusCode: '200',
          },
          type: 'fixed-response',
        },
      ],
      loadBalancerArn: uiAlb.loadBalancerArn,
      port: 443,
      protocol: 'HTTPS',
      certificates: [{ certificateArn: props.certificateArn }],
    });

    // const action = new CfnListener.ActionProperty();
    const internalListener = new CfnListener(this, 'InternalListener', {
      defaultActions: [
        {
          fixedResponseConfig: {
            contentType: 'text/plain',
            messageBody: 'hi from the default last rule',
            statusCode: '200',
          },
          type: 'fixed-response',
        },
      ],
      loadBalancerArn: uiAlb.loadBalancerArn,
      port: 9443,
      protocol: 'HTTPS',
      certificates: [{ certificateArn: props.certificateArn }],
    });

    new CfnOutput(this, 'ListenerOutput', {
      value: listener.ref,
      exportName: `${props.envName}-Listener`,
    });

    this.listener = listener;

    new CfnOutput(this, 'InternalListenerOutput', {
      value: internalListener.ref,
      exportName: `${props.envName}-InternalListener`,
    });

    this.internalListener = internalListener;

    this.albDnsAddress = uiAlb.loadBalancerDnsName;
  }
}

export class AppNlbComboInterfaceVpcEndpoint extends Construct {
  public readonly interfaceVpcEndpoint: IInterfaceVpcEndpoint;
  public readonly vpcEndpointId: string;
  public readonly endpointDns: string;

  constructor(scope: Construct, id: string, props: AppNlbComboInterfaceVpcEndpointProps) {
    super(scope, id);

    // default is to allow root account to use endpoint
    const whitelistedPrincipals = props.whitelistedPrincipals || [
      new ArnPrincipal(Fn.sub('arn:aws:iam::${AWS::AccountId}:root')),
    ];

    const vpcEndpointService = new VpcEndpointService(this, 'vpcEndpointService', {
      acceptanceRequired: false,
      vpcEndpointServiceLoadBalancers: props.vpcEndpointServiceLoadBalancers,
      whitelistedPrincipals,
    });

    const cfnVPCEndpointService = vpcEndpointService.node.findChild('vpcEndpointService') as CfnVPCEndpointService;

    const subnetType = props.subnetType || SubnetType.PRIVATE;
    const port = props.port || 443;

    this.interfaceVpcEndpoint = new InterfaceVpcEndpoint(this, 'endpoint', {
      vpc: props.vpc,
      service: {
        name: Fn.sub('com.amazonaws.vpce.${AWS::Region}.${serviceName}', {
          serviceName: cfnVPCEndpointService.ref,
        }),
        port,
      },
      privateDnsEnabled: false,
      subnets: {
        subnetType,
      },
    });

    const exportName = props.exportName || 'NlbVpcEndpointId';

    new CfnOutput(this, 'NlbVpcEndpointId', {
      value: this.interfaceVpcEndpoint.vpcEndpointId,
      exportName: `${props.envName}-${exportName}`,
    });

    this.vpcEndpointId = this.interfaceVpcEndpoint.vpcEndpointId;

    const cfnVpcEndpoint = this.interfaceVpcEndpoint.node.findChild('Resource') as CfnVPCEndpoint;

    new CfnOutput(this, 'ListenerOutput', {
      value: Fn.select(1, Fn.split(':', Fn.select(0, cfnVpcEndpoint.attrDnsEntries))),
      exportName: `${props.envName}-EndpointDns`,
    });

    this.endpointDns = Fn.select(1, Fn.split(':', Fn.select(0, cfnVpcEndpoint.attrDnsEntries)));
  }
}

export class NlbEndpointCombo extends Construct {
  constructor(scope: Construct, id: string, props: NlbEndpointComboProps) {
    super(scope, id);

    const vpc = props.appAlbNlbVpc;

    // Default is to use same vpc
    let remoteEndpointVpc = props.appAlbNlbVpc;

    if (props.nlbEndpointVpc !== undefined) {
      remoteEndpointVpc = props.nlbEndpointVpc;
    }

    // Network Load Balancers
    // TODO: subnets, sg
    const netAlb = new NetworkLoadBalancer(this, `${props.envName}-nlb`, {
      vpc,
      // TODO: props
      internetFacing: false,
      loadBalancerName: `${props.envName}-nlb`,
      deletionProtection: false,
      // TODO: props
      crossZoneEnabled: true,
      // vpcSubnets: subnets,
      // securityGroup: albSecurityGroup,
    });

    // AccessLogs only for TLS
    // const logBucket = s3.Bucket.fromBucketName(this, 'logBucket', props.logBucket);
    // netAlb.logAccessLogs(logBucket, props.bucketPrefix);

    const networkTargetGroup = new NetworkTargetGroup(this, `${props.envName}-nlb-tg`, {
      port: 443,
      vpc,
      targetType: TargetType.IP,
    });

    new NetworkListener(this, `${props.envName}-nlb-listener`, {
      loadBalancer: netAlb,
      port: 443,
      defaultTargetGroups: [networkTargetGroup],
    });

    const parameterPath = props.parameterPath || '/App/Alb/app-nlb/';
    const parameterName = `${parameterPath}${props.envName}/dns-saved`;

    // Create a new SSM Parameter holding a String
    new StringParameter(this, 'StringParameter', {
      parameterName,
      stringValue: 'x',
    });

    const inputText = {
      alb_dns_address: props.albDnsAddress,
      alb_dns_parameter: parameterName,
      debug: true,
      dry_run: false,
      alb_target_group_arn: networkTargetGroup.targetGroupArn,
    };

    new AppNlbComboCloudwatchLambdaRule(this, 'appAlbNlbRule', {
      lambdaForUpdatingTargets: props.lambdaForUpdatingTargets,
      // targetInputText: JSON.stringify(inputText),
      targetInputObject: inputText,
    });

    const appNlbComboInterfaceVpcEndpoint = new AppNlbComboInterfaceVpcEndpoint(this, 'appNlbEndpoint', {
      vpc: remoteEndpointVpc,
      envName: props.envName,
      vpcEndpointServiceLoadBalancers: [netAlb],
    });

    // If you define a zone we can publish the dns - otherwise leave it up to the app to do this bit
    if ((props.publicHostedZoneId && !props.zoneName) || (!props.publicHostedZoneId && props.zoneName)) {
      throw new Error('both publicHostedZoneId and zoneName are needed or else both should be empty');
    } else if (props.publicHostedZoneId !== undefined && props.zoneName !== undefined) {
      // TODO: move this back to using constructs
      // prettier-ignore
      // eslint-disable-next-line
      const cfnVpcEndpoint = appNlbComboInterfaceVpcEndpoint.interfaceVpcEndpoint.node.findChild('Resource') as CfnVPCEndpoint;
      // prettier-ignore-end

      const recordSetPrefix = props.recordSetPrefix || 'app-alb-nlb-combo';

      new CfnRecordSet(this, 'EndpointRecordSet', {
        hostedZoneId: props.publicHostedZoneId,
        // FIXME: move to props as string not as import
        name: Fn.join('.', [recordSetPrefix, props.zoneName]),
        type: 'CNAME',
        resourceRecords: [Fn.select(1, Fn.split(':', Fn.select(0, cfnVpcEndpoint.attrDnsEntries)))],
        ttl: '60',
        // Comment, record_set[:comment]
      });

      // FIXTO:
      // new CnameRecord(this, 'cname', {
      //   domainName: endpointdns,
      //   recordName: props.baseHost,
      //   zone: zone
      // }
    }
  }
}
