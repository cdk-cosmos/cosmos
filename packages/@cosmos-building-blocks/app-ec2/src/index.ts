// export {

// } from './app-alb-nlb-combo';

export * from './app-alb-nlb';
