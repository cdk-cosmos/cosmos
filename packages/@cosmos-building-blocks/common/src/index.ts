export { CDKToolkit } from './cdktoolkit';
export { CrossAccountExports, CrossAccountExportsFn, CrossAccountExportsProps } from './cross-account-exports';
