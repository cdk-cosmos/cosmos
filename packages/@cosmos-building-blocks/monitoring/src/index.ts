export { ecsMetric } from './helpers';
export { ClusterDashboard, ClusterDashboardProps } from './dashboard/cluster';
