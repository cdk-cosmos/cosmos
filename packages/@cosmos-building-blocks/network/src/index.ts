export { IResolverRule, ResolverRule, ResolverRuleProps } from './route53/resolver-rule';
export {
  IResolverRuleAssociation,
  ResolverRuleAssociation,
  ResolverRuleAssociationProps,
} from './route53/resolver-rule-association';

export { ITransitGateway, TransitGateway, TransitGatewayProps } from './transit-gateway/transit-gateway';
export {
  ITransitGatewayAttachment,
  TransitGatewayAttachment,
  TransitGatewayAttachmentProps,
  TransitGatewayRoute,
  TransitGatewayRouteProps,
} from './transit-gateway/transit-gateway-attachment';

export { TransparentProxy, TransparentProxyProps } from './internet/transparent-proxy';
