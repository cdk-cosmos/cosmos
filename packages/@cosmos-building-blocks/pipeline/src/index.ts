export { AppNodePipeline, AppNodePipelineProps } from './app-node-pipeline';
