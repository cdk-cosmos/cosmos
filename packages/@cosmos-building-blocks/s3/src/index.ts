// export {

// } from './app-alb-nlb-combo';

export * from './s3';
