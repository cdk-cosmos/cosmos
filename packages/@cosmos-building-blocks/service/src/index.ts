export { EcsService, EcsServiceProps } from './ecs';
