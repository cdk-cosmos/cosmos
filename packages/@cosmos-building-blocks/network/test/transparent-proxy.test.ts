import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { App, Stack } from '@aws-cdk/core';
import { Vpc, SubnetType } from '@aws-cdk/aws-ec2';
import { TransparentProxy } from '../src';

const app = new App();
const stack = new Stack(app, 'Test');
const vpc = new Vpc(stack, 'VPC', {
  cidr: '10.180.7.0/24',
  maxAzs: 2,
  subnetConfiguration: [
    {
      name: 'Main',
      subnetType: SubnetType.ISOLATED,
      cidrMask: 26,
    },
    {
      name: 'Redis',
      subnetType: SubnetType.ISOLATED,
      cidrMask: 28,
    },
  ],
});

new TransparentProxy(stack, 'TransparentProxy', {
  vpc: vpc,
  vpcSubnets: { subnetGroupName: 'Main' },
  host: 'proxy.internal',
  port: 8080,
  initialInstanceId: 'in-1234',
});

const synth = SynthUtils.synthesize(stack);

describe('Transparent Proxy', () => {
  test('should match snapshot', () => {
    expect(synth.template).toMatchSnapshot();
  });
});
