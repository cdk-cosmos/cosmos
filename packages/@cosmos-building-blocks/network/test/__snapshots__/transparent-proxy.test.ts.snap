// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transparent Proxy should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "SsmParameterValueawsserviceamiamazonlinuxlatestamzn2amihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter": Object {
      "Default": "/aws/service/ami-amazon-linux-latest/amzn2-ami-hvm-x86_64-gp2",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": Object {
    "TransparentProxyAsgASGD26968BB": Object {
      "Properties": Object {
        "DesiredCapacity": "1",
        "LaunchConfigurationName": Object {
          "Ref": "TransparentProxyAsgLaunchConfig18394516",
        },
        "MaxSize": "1",
        "MinSize": "1",
        "Tags": Array [
          Object {
            "Key": "Name",
            "PropagateAtLaunch": true,
            "Value": "Test/TransparentProxy/Asg",
          },
        ],
        "VPCZoneIdentifier": Array [
          Object {
            "Ref": "VPCMainSubnet1SubnetE776C424",
          },
          Object {
            "Ref": "VPCMainSubnet2Subnet1B273166",
          },
        ],
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
      "UpdatePolicy": Object {
        "AutoScalingScheduledAction": Object {
          "IgnoreUnmodifiedGroupSizeProperties": true,
        },
      },
    },
    "TransparentProxyAsgInstanceProfileE888F303": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "TransparentProxyAsgInstanceRole97CEB8B8",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "TransparentProxyAsgInstanceRole97CEB8B8": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMFullAccess",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test/TransparentProxy/Asg",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TransparentProxyAsgInstanceRoleDefaultPolicy3847ADE0": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ec2:ModifyInstanceAttribute",
                "ec2:DescribeInstances",
                "ec2:DescribeRouteTables",
                "ec2:ReplaceRoute",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TransparentProxyAsgInstanceRoleDefaultPolicy3847ADE0",
        "Roles": Array [
          Object {
            "Ref": "TransparentProxyAsgInstanceRole97CEB8B8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TransparentProxyAsgInstanceSecurityGroupC07B4566": Object {
      "Properties": Object {
        "GroupDescription": "Test/TransparentProxy/Asg/InstanceSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "from 0.0.0.0/0:ALL TRAFFIC",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test/TransparentProxy/Asg",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "TransparentProxyAsgLaunchConfig18394516": Object {
      "DependsOn": Array [
        "TransparentProxyAsgInstanceRoleDefaultPolicy3847ADE0",
        "TransparentProxyAsgInstanceRole97CEB8B8",
      ],
      "Properties": Object {
        "IamInstanceProfile": Object {
          "Ref": "TransparentProxyAsgInstanceProfileE888F303",
        },
        "ImageId": Object {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestamzn2amihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t3.nano",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "TransparentProxyAsgInstanceSecurityGroupC07B4566",
              "GroupId",
            ],
          },
        ],
        "UserData": Object {
          "Fn::Base64": Object {
            "Fn::Join": Array [
              "",
              Array [
                "#!/bin/bash
exec > >(tee /var/log/user-data.log|logger -t user-data -s 2>/dev/console) 2>&1
export HTTP_PROXY=http://proxy.internal:8080 && export HTTPS_PROXY=$HTTP_PROXY
yum install -y squid iptables-services
unset HTTP_PROXY && unset HTTPS_PROXY
iptables -t nat -A PREROUTING -p tcp --dport 80 -j REDIRECT --to-port 8081
iptables -t nat -A PREROUTING -p tcp --dport 443 -j REDIRECT --to-port 8082
iptables-save > /etc/sysconfig/iptables
systemctl enable squid
systemctl enable iptables
mkdir /etc/squid/ssl
cd /etc/squid/ssl
openssl genrsa -out squid.key 4096
openssl req -new -key squid.key -out squid.csr -subj \\"/L=proxy/O=cosmos/CN=squid\\"
openssl x509 -req -days 3650 -in squid.csr -signkey squid.key -out squid.crt
cat squid.key squid.crt >> squid.pem
cd /etc/squid
echo http_port 8080 >> squid.conf
echo http_port 8081 intercept >> squid.conf
echo https_port 8082 cert=/etc/squid/ssl/squid.pem ssl-bump intercept >> squid.conf
echo ssl_bump splice all >> squid.conf
echo cache_peer proxy.internal parent 8080 0 default >> squid.conf
echo never_direct allow all >> squid.conf
echo http_access allow all >> squid.conf
systemctl start squid
systemctl start iptables
region=",
                Object {
                  "Ref": "AWS::Region",
                },
                "
instanceId=$(curl -s http://169.254.169.254/latest/meta-data/instance-id)
aws ec2 modify-instance-attribute --region=\${region} --instance-id=\${instanceId} --no-source-dest-check
declare -a routingTableIds=(",
                Object {
                  "Ref": "VPCMainSubnet1RouteTableFDDE4665",
                },
                " ",
                Object {
                  "Ref": "VPCMainSubnet2RouteTableA47F3294",
                },
                ")
for routingTableId in \\"\${routingTableIds[@]}\\"
do
aws ec2 replace-route --region=\${region} --route-table-id=\${routingTableId} --destination-cidr-block=0.0.0.0/0 --instance-id=\${instanceId}
done",
              ],
            ],
          },
        },
      },
      "Type": "AWS::AutoScaling::LaunchConfiguration",
    },
    "TransparentProxyMainSubnet1DefaultRoute416C1450": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "InstanceId": "in-1234",
        "RouteTableId": Object {
          "Ref": "VPCMainSubnet1RouteTableFDDE4665",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TransparentProxyMainSubnet2DefaultRoute87B94861": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "InstanceId": "in-1234",
        "RouteTableId": Object {
          "Ref": "VPCMainSubnet2RouteTableA47F3294",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCB9E5F0B4": Object {
      "Properties": Object {
        "CidrBlock": "10.180.7.0/24",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test/VPC",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VPCMainSubnet1RouteTableAssociationA92E4DDA": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCMainSubnet1RouteTableFDDE4665",
        },
        "SubnetId": Object {
          "Ref": "VPCMainSubnet1SubnetE776C424",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCMainSubnet1RouteTableFDDE4665": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test/VPC/MainSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCMainSubnet1SubnetE776C424": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.180.7.0/26",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test/VPC/MainSubnet1",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Main",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCMainSubnet2RouteTableA47F3294": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test/VPC/MainSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCMainSubnet2RouteTableAssociation81BDC51F": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCMainSubnet2RouteTableA47F3294",
        },
        "SubnetId": Object {
          "Ref": "VPCMainSubnet2Subnet1B273166",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCMainSubnet2Subnet1B273166": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.180.7.64/26",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test/VPC/MainSubnet2",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Main",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCRedisSubnet1RouteTableAssociation1C673EBE": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCRedisSubnet1RouteTableCAF7FB26",
        },
        "SubnetId": Object {
          "Ref": "VPCRedisSubnet1SubnetA76F661B",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCRedisSubnet1RouteTableCAF7FB26": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test/VPC/RedisSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCRedisSubnet1SubnetA76F661B": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.180.7.128/28",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test/VPC/RedisSubnet1",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Redis",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCRedisSubnet2RouteTable16EB92BE": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test/VPC/RedisSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCRedisSubnet2RouteTableAssociationE7F24ECB": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCRedisSubnet2RouteTable16EB92BE",
        },
        "SubnetId": Object {
          "Ref": "VPCRedisSubnet2SubnetB4A3AE92",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCRedisSubnet2SubnetB4A3AE92": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.180.7.144/28",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test/VPC/RedisSubnet2",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Redis",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
  },
}
`;
